### PR TITLE
Use more C++11 features:

### PR DESCRIPTION
**Pull request by nbougalis**
* Remove beast::static_initializer
* Remove noexcept VS2013 workaround
* Use variadic constructor
* Use [[noreturn]] attribute
* Default move constructors

### CODE REVIEW

**Comment by HowardHinnant** [comment]
Just a few minor suggested tweaks.  I love how much smaller and simpler this makes rippled!

File: src/beast/beast/threads/SharedData.h [hunk 1]
```diff
@@ -125,45 +125,16 @@ class SharedData
...
-
-    template <class T1, class T2, class T3, class T4, class T5, class T6>
-    SharedData (T1 t1, T2 t2, T3 t3, T4 t4, T5 t5, T6 t6)
-        : m_value (t1, t2, t3, t4, t5, t6) { }
-
-    template <class T1, class T2, class T3, class T4, class T5, class T6, class T7>
-    SharedData (T1 t1, T2 t2, T3 t3, T4 t4, T5 t5, T6 t6, T7 t7) : m_value (t1, t2, t3, t4, t5, t6, t7) { }
+    template <class... Params>
+    SharedData (Params... params)
+        : m_value (std::forward<Params>(params)...)
```
Comment (seelabs) [inline src/beast/beast/threads/SharedData.h]: Make the constructor `explicit` and `Params` should be forwarding refs (i.e. `&&`)

File: src/ripple/rpc/handlers/AccountObjects.cpp [hunk 2]
```diff
@@ -81,7 +81,7 @@ Json::Value doAccountObjects (RPC::Context& context)
     {
         using filter_types = std::vector <std::pair <std::string, LedgerEntryType>>;
         static
-        beast::static_initializer <filter_types> const types ([]() -> filter_types {
+        filter_types const types ([]() -> filter_types {
```
Comment (seelabs) [inline src/ripple/rpc/handlers/AccountObjects.cpp]: Not part of this PR, but I don't see why there's a lambda here at all. It can be removed.

**Comment by seelabs** [comment]
@nbougalis: Windows appears to be failing unit tests (see appveyor): `Assertion failed: type_ == nullValue || type_ == arrayValue, file C:\projects\rippled-179\src\ripple/json/impl/json_value.cpp, line 819`

**Comment by seelabs** [comment]
These changes clean things up nicely. I'll give a thumbs up after the issues with windows unittests pass and the `SharedData` constructor are resolved.

**Comment by nbougalis** [comment]
@seelabs: I saw that failure and am not sure what's going on there.

**Comment by ximinez** [comment]
@nbougalis @seelabs Appveyor has a 40 minute time limit, and fails silently when it hits it. Seems to be more related to Appveyor being busy and not anything in our code. Unfortunately, I'm not able to restart the build.

File: src/beast/beast/threads/SharedData.h [hunk 3]
```diff
@@ -125,45 +126,18 @@ class SharedData
...
-        : m_value (t1, t2, t3, t4, t5) { }
-
-    template <class T1, class T2, class T3, class T4, class T5, class T6>
-    SharedData (T1 t1, T2 t2, T3 t3, T4 t4, T5 t5, T6 t6)
-        : m_value (t1, t2, t3, t4, t5, t6) { }
-
-    template <class T1, class T2, class T3, class T4, class T5, class T6, class T7>
-    SharedData (T1 t1, T2 t2, T3 t3, T4 t4, T5 t5, T6 t6, T7 t7) : m_value (t1, t2, t3, t4, t5, t6, t7) { }
+    template <class... Params,
+        class = std::enable_if_t<std::is_constructible<Value, Params...>{}>
```
Comment (seelabs) [inline src/beast/beast/threads/SharedData.h]: Shouldn't this be type `forward` type? (what if a type is constructable from an rvalue ref only)

Comment (HowardHinnant) [inline src/beast/beast/threads/SharedData.h]: Nope.
If the argument is an lvalue `T`, `Params[i]` will be `T&`, and we ask the question: `is_constructible<Value, T&>` (is constructible from an lvalue `T`.
If the argument is an rvalue `T`, `Params[i]` will be `T`, and we ask the question: `is_constructible<Value, T>` (is constructible from an rvalue `T`.
In other words, it just works.

Comment (seelabs) [inline src/beast/beast/threads/SharedData.h]: Interesting! So even though the declaration `T a` in the function is a value type, in the template parameter it's a ref? TIL.

Comment (HowardHinnant) [inline src/beast/beast/threads/SharedData.h]: Not quite.  The parameter `P&&` becomes `T&&` when bound to an rvalue `T`.  In this case `P` deduced to `T`.  If instead `P&&` binds to an lvalue `T`, it becomes `T&`.  In this case `P` deduced to `T&`, and `T& &&` collapsed to `T&`.

**Comment by HowardHinnant** [comment]
This all tested out for me. :+1:   I haven't looked at the travis/appveyor failures.

**Comment by nbougalis** [comment]
Well, VS2015 isn't cooperating:

    C:\projects\rippled-179\src\ripple/app/ledger/impl/LedgerCleaner.cpp(98): error C2512: 'beast::SharedData<ripple::LedgerCleanerImp::State,beast::SharedMutexAdapter<beast::RecursiveMutex>>::SharedData': no appropriate default constructor available 
    scons: *** [build\msvc.debug\src\ripple\unity\app_ledger.obj] Error 2
    scons: building terminated because of errors.

File: src/beast/beast/threads/SharedData.h (hunk 3 above)
Comment (seelabs) [inline src/beast/beast/threads/SharedData.h]: @HowardHinnant Code example: http://melpon.org/wandbox/permlink/C4NIYlsYs84UlBAj 

Above you said we ask the question: `is_constructable<V, T&>. I still think we are asking the question `is_constructable<V, T>`, where T is a *value* type, not a reference type. The reason the code above works, is `is_constructable<V, T>` is true even if V is only constructible from an rvalue ref and not an lvalue ref.

(To be clear: I agree the code as written is correct, I'm quibbling with the explanation as to why it's correct because if I'm wrong I need to educate myself).

Comment (HowardHinnant) [inline src/beast/beast/threads/SharedData.h]: You are correct.  As far as this trait is concerned, `T` and `T&&` (in an argument position -- 2nd or more) are identical in meaning and both mean "rvalue".  In general, sometimes the traits mean "expression", but "type" is the only language they speak.  And expressions never have reference type.  Instead they are lvalue or rvalue.  So when we need to express an "lvalue expression", that is approximated with an lvalue reference type (e.g. `T&`).  And when we need to express an "rvalue expression", that is approximated with both `T` and `T&&`.

And just in case you're not confused enough, `is_constructible<Value, Params&&...>` would also work here (and mean exactly the same thing).  If `Param[i]` is `T&`, then we have `T& &&`, which collapses to `T&`, so `is_constructible` sees an lvalue arg.  And if `Param[i]` is `T`, then we have `T&&` and `is_constructible` sees an rvalue arg.

**Comment by vinniefalco** [comment]
I prefer we removed `SharedData` instead of "improving" it.

**Comment by nbougalis** [comment]
@vinniefalco as a matter of fact: https://github.com/nbougalis/rippled/commits/shareddata

**Comment by seelabs** [comment]
@nbougalis It looks like VS2015 is not handling the `is_constructible` correctly in `SharedData.h` (or less likely, the `enable_if`). This patch removes the `enable_if`: https://github.com/seelabs/rippled/commit/2976c0470b96c8978d08719cd23f6fb729e7cbbc

**Comment by vinniefalco** [comment]
Better to NOT TOUCH the class and REMOVE It instead!

**Comment by nbougalis** [comment]
I have removed the changes to `SharedDate` since the entire class is removed with #1348. All other review comments and suggestions have been incorporated and the changes have been squashed.

**Comment by seelabs** [comment]
:+1: Side note: Gcc 5.2 is issuing two warning about unused variables: 

1) CreateOffer.cpp:652:17: warning: unused variable ‘uPaysCurrency’ 

2) Transactor.cpp:665:23: warning: unused variable ‘a_seq’

However, these are unrelated to your changes.

**Comment by ximinez** [comment]
Those warnings are my fault, and are fixed in https://github.com/ximinez/rippled/commit/08ff164ca6db35e5e45d2497af401e41b14ed3ac (part of #1230)

**Comment by nbougalis** [comment]
Merged as 570bb2e13920e386f10f1295cf1e62ca41fce05d.
